pkg/controlplane/control/manager.go

```diff
@@ -469,46 +614,235 @@ func (m *Manager) deleteImportService(ctx context.Context, service, imp types.Na
+		if !k8serrors.IsNotFound(err) {
+			return err
+		}
+
+		m.logger.Infof(
+			"Deleting an import endpoint slice with no corresponding dataplane endpoint slice: %s",
+			endpointSliceName)
+		shouldDelete = true
 	}
 
-	if name, ok := service.Labels[LabelImportName]; !ok || name != importName.Name {
-		return conflictingServiceError{
-			name:      serviceName,
-			managedBy: managedBy,
+	if shouldDelete {
+		err := m.client.Delete(ctx, &discv1.EndpointSlice{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      endpointSliceName,
+				Namespace: namespace,
+			},
+		})
+		if err != nil && !k8serrors.IsNotFound(err) {
+			return err
 		}
+
+		return nil
 	}
 
-	if namespace, ok := service.Labels[LabelImportNamespace]; !ok || namespace != importName.Namespace {
-		return conflictingServiceError{
-			name:      serviceName,
-			managedBy: managedBy,
+	return m.checkImportEndpointSlice(ctx, &imp, &dataplaneEndpointSlice)
+}
+
+func (m *Manager) checkImportEndpointSlice(
+	ctx context.Context,
+	imp *v1alpha1.Import,
+	dataplaneEndpointSlice *discv1.EndpointSlice,
+) error {
+	m.logger.Infof(
+		"Checking endpoint slice %s for import %s/%s'.",
+		dataplaneEndpointSlice.Name, imp.Namespace, imp.Name)
+
+	importEndpointSliceName := (&importEndpointSliceName{
+		importName:                 imp.Name,
+		dataplaneEndpointSliceName: dataplaneEndpointSlice.Name,
+	}).Get()
+	protocol := v1.ProtocolTCP
+	port32 := int32(imp.Spec.TargetPort)
+
+	importEndpointSlice := discv1.EndpointSlice{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      importEndpointSliceName,
+			Namespace: imp.Namespace,
+			Labels: map[string]string{
+				discv1.LabelServiceName:  imp.Name,
+				discv1.LabelManagedBy:    AppName,
+				LabelDPEndpointSliceName: dataplaneEndpointSlice.Name,
+			},
+		},
+		AddressType: discv1.AddressTypeIPv4,
+		Endpoints:   dataplaneEndpointSlice.Endpoints,
+		Ports: []discv1.EndpointPort{
+			{
+				Port:     &port32,
+				Protocol: &protocol,
+			},
+		},
+	}
+
+	var oldImportEndpointSlice discv1.EndpointSlice
+	err := m.client.Get(
+		ctx,
+		types.NamespacedName{
+			Name:      importEndpointSliceName,
+			Namespace: imp.Namespace,
+		},
+		&oldImportEndpointSlice)
+	if err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return err
+		}
+
+		m.logger.Infof("Creating import endpoint slice: %s.", importEndpointSliceName)
+		return m.client.Create(ctx, &importEndpointSlice)
+	}
+
+	if !endpointSliceChanged(&importEndpointSlice, &oldImportEndpointSlice) {
+		return nil
+	}
+
+	m.logger.Infof("Updating import endpoint slice: %s.", importEndpointSliceName)
+	return m.client.Update(ctx, &importEndpointSlice)
+}
+
+func (m *Manager) addImportEndpointSlices(ctx context.Context, imp *v1alpha1.Import) error {
+	// check that import service exists
+	importName := types.NamespacedName{
+		Namespace: imp.Namespace,
+		Name:      imp.Name,
+	}
+
+	if err := m.client.Get(ctx, importName, &v1.Service{}); err != nil {
+		if k8serrors.IsNotFound(err) {
```

Comment:
The specific scenario we were targeting endpointslice involves, the service could to be created externally by another application asynchronously. Hence, it might be a strong assumption to assume the service should already be present esp in Non-CRD mode.